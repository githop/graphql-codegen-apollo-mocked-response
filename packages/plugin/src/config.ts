import { RawConfig } from '@graphql-codegen/visitor-plugin-common';

export type Config = Pick<
  RawConfig,
  'namingConvention' | 'typesPrefix' | 'typesSuffix'
> & {
  typesFile: string;
  addTypename: boolean;
  prefix?: string;
};

export function getConfig(config?: Partial<Config>): Config {
  const baseConfig = new Map<string, any>([
    ['typesFile', './types.ts'],
    ['addTypename', false],
    ['namingConvention', 'change-case-all#pascalCase'],
    ['typesPrefix', null],
    ['typesSuffix', null],
  ]);

  const configMap = new Map<string, any>(Object.entries(config ?? {}));

  const resolved = new Map([...baseConfig, ...configMap]);
  // drop extension
  resolved.set('typesFile', resolved.get('typesFile')!.replace(/\.[\w]+$/, ''));
  return Object.fromEntries(resolved) as Config;
}
