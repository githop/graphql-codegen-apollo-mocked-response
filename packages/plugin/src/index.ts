import { PluginFunction } from '@graphql-codegen/plugin-helpers';
import {
  concatAST,
  TypeInfo,
  GraphQLCompositeType,
  visitWithTypeInfo,
  visit,
  GraphQLOutputType,
  TypeNode,
  OperationDefinitionNode,
  DefinitionNode,
} from 'graphql';
import a from 'indefinite';

const APOLLO_IMPORTS = [
  "import { ApolloError } from '@apollo/client';",
  "import { MockedResponse } from '@apollo/client/testing';",
];

const MOCK_FN_INTERFACE = `
  interface MockFn<Variables, Query> {
    ({
      variables,
      result,
      error,
    }: {
      variables?: Variables;
      result?: Query;
      error?: ApolloError;
    }): MockedResponse<Query>;
  }  
  `;

function hasListType(typeNode: TypeNode): boolean {
  if (typeNode.kind === 'ListType') {
    return true;
  }

  if (typeNode.kind === 'NamedType') {
    return false;
  }
  return hasListType(typeNode.type);
}

function buildQueryResult(
  transformedSelections: TransformedField[],
  config: Config,
  types = new Set<string>()
) {
  let ret = '';
  for (const node of transformedSelections) {
    if (node == transformedSelections[0] && config.addTypename) {
      ret += ` __typename: '${node.parentType}',`;
    }
    if (node.selections == null) {
      types.add(String(node.parentType));
      const mockInstance = mockInstanceName(`${node.parentType}`);
      ret += ` ${node.fieldName}: ${mockInstance}.${node.fieldName},`;
    } else {
      const [fields] = buildQueryResult(node.selections, config, types);
      const embed =
        node.kind === 'connection' ? `[{${fields}}]` : `{ ${fields} }`;
      ret += ` ${node.fieldName}: ${embed}, `;
    }
  }
  return [ret, [...types]] as const;
}

interface BuildMockFn {
  queryResultType: string;
  queryVarsType: string;
  functionName: string;
  documentNode: string;
  queryResult: string;
  queryResultVar: string;
  mocks: string;
}

function mockFnTemplate({
  functionName,
  queryVarsType,
  queryResultType,
  documentNode,
  queryResult,
  queryResultVar,
  mocks,
}: BuildMockFn) {
  return `
  export const ${functionName}: MockFn<
  ${queryVarsType},
  ${queryResultType}
> = ({ result, variables, error }) => {
  ${mocks}

  ${queryResult}

  return {
    request: {
      query: ${documentNode},
      variables,
    },
    result: {
      data: result != null ? result : ${queryResultVar},
      error,
    },
  };
};
  `;
}

function mockFnName(type: string, prefix?: string) {
  if (prefix) {
    return `${prefix}${type}`;
  }
  return a(type, { articleOnly: true }) + type;
}

function mockInstanceName(type: string) {
  return type[0].toLowerCase() + type.substring(1) + 'Mock';
}

interface TransformedField {
  parentType: GraphQLCompositeType;
  type: GraphQLOutputType;
  fieldName: string;
  // primitive are scalars and connections are just embedded objects
  kind: 'primitive' | 'connection';
  selections?: Array<TransformedField>;
}

function handleOperationResult(
  operation: OperationDefinitionNode,
  config: Config
) {
  const functionName = `mock${operation.name?.value}`;
  const queryResultType = `${operation.name?.value}Query`;
  const queryVarsType = `${operation.name?.value}QueryVariables`;
  const documentNode = `${operation.name?.value}Document`;

  const transformedFields =
    (operation.selectionSet.selections as unknown as TransformedField[]) ?? [];

  const [resultFields, types] = buildQueryResult(transformedFields, config);

  const mocks = types
    .map((typeStr) => {
      const mockFn = mockFnName(typeStr, config.prefix);
      const mockInstance = mockInstanceName(typeStr);
      return `const ${mockInstance} = ${mockFn}();`;
    })
    .join('\n');

  const queryResultVar = `${operation.name?.value}Result`;
  const queryResult = `const ${queryResultVar}: ${queryResultType} = {${resultFields}}`;
  const templateVars = {
    functionName,
    queryResultType,
    queryVarsType,
    mocks,
    queryResult,
    documentNode,
    queryResultVar,
  };

  const imports = [documentNode, queryResultType, queryVarsType];

  return {
    templateVars,
    imports,
  };
}

export interface Config {
  typesFile: string;
  addTypename: boolean;
  prefix?: string;
}

export function getConfig(config?: Partial<Config>): Config {
  const baseConfig: Config = {
    typesFile: './types.ts',
    addTypename: false,
  };
  const resolved = Object.assign(baseConfig, config);
  // drop extension
  resolved.typesFile = resolved.typesFile.replace(/\.[\w]+$/, '');
  return resolved;
}

export const plugin: PluginFunction<Partial<Config>> = (
  schema,
  documents,
  config
) => {
  const docsAST = concatAST(documents.map((v) => v.document!));
  const typeInfo = new TypeInfo(schema);

  const result = visit(
    docsAST,
    visitWithTypeInfo(typeInfo, {
      Field: {
        leave(node) {
          const parentType = typeInfo.getParentType()!;
          const type = typeInfo.getType()!;

          const transformed: TransformedField = {
            parentType,
            type,
            fieldName: node.name.value,
            selections: node.selectionSet?.selections as any[],
            kind: hasListType(typeInfo.getFieldDef()!.astNode!.type)
              ? 'connection'
              : 'primitive',
          };
          return transformed;
        },
      },
    })
  );

  const resolvedConfig = getConfig(config);

  function isOperationDefinitionPredicate(
    node: DefinitionNode,
    index: number,
    array: readonly DefinitionNode[]
  ): node is OperationDefinitionNode {
    return 'kind' in node && node.kind === 'OperationDefinition';
  }

  const templateVars = result.definitions
    .filter(isOperationDefinitionPredicate)
    .map((operation) => handleOperationResult(operation, resolvedConfig));

  const typeImports = templateVars.flatMap(({ imports }) => imports);

  const imports = [
    ...APOLLO_IMPORTS,
    `import { ${typeImports.join(', ')} } from '${resolvedConfig.typesFile}'`,
  ];

  const lines: string[] = [MOCK_FN_INTERFACE];

  return {
    prepend: imports,
    content: lines
      .concat(
        templateVars.map(({ templateVars }) => mockFnTemplate(templateVars))
      )
      .join('\n'),
  };
};
