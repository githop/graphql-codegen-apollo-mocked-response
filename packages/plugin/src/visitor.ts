import {
  TypeInfo,
  visitWithTypeInfo,
  visit,
  DocumentNode,
  GraphQLSchema,
  FragmentSpreadNode,
  TypeNode,
  FragmentDefinitionNode,
  Kind,
  SelectionSetNode,
  OperationDefinitionNode,
} from 'graphql';

export interface TransformedField {
  parentTypename: string;
  fieldName: string;
  // primitive are scalars and connections are just embedded objects
  kind: 'primitive' | 'connection';
  selections?: TransformedSelections;
}

export type TransformedSelections = Array<
  TransformedField | TransformedFragmentSpread
>;

export interface TransformedFragmentSpread {
  name: string;
  kind: FragmentSpreadNode['kind'];
}

export interface TransformedFragment {
  name: string;
  kind: FragmentDefinitionNode['kind'];
  type: string;
  selections: TransformedSelections;
}

export interface TransformedSelectionSet {
  kind: SelectionSetNode['kind'];
  selections: TransformedSelections;
}

export interface TransformedOperation {
  kind: OperationDefinitionNode['kind'];
  operation: OperationDefinitionNode['operation'];
  selectionSet: TransformedSelectionSet;
  name: string;
}

export function isTransformedField(node: any): node is TransformedField {
  return (
    ('kind' in node && node.kind === 'primitive') || node.kind === 'connection'
  );
}

export function isTransformedFragment(node: any): node is TransformedFragment {
  return (
    'kind' in node &&
    node.kind === Kind.FRAGMENT_DEFINITION &&
    'selections' in node
  );
}

export function isTransformedOperation(
  node: any
): node is TransformedOperation {
  return (
    'kind' in node &&
    node.kind === Kind.OPERATION_DEFINITION &&
    'name' in node &&
    typeof node.name === 'string'
  );
}

export function hasListType(typeNode: TypeNode): boolean {
  if (typeNode.kind === 'ListType') {
    return true;
  }

  if (typeNode.kind === 'NamedType') {
    return false;
  }
  return hasListType(typeNode.type);
}

export function visitor(ast: DocumentNode, schema: GraphQLSchema) {
  const typeInfo = new TypeInfo(schema);

  return visit(
    ast,
    visitWithTypeInfo(typeInfo, {
      FragmentSpread: {
        leave(node) {
          const transformedFragmentSpread: TransformedFragmentSpread = {
            kind: node.kind,
            name: node.name as any,
          };
          return transformedFragmentSpread;
        },
      },
      FragmentDefinition: {
        leave(node) {
          const transformedFragment: TransformedFragment = {
            name: node.name as any,
            kind: node.kind,
            type: node.typeCondition.name as any,
            selections: node.selectionSet.selections as any,
          };
          return transformedFragment;
        },
      },
      OperationDefinition: {
        leave(node) {
          const transformedOperation: TransformedOperation = {
            kind: node.kind,
            operation: node.operation,
            name: node.name as any,
            selectionSet: node.selectionSet as any,
          };
          return transformedOperation;
        },
      },
      SelectionSet: {
        leave(node) {
          const transformedSelectionSet: TransformedSelectionSet = {
            kind: node.kind,
            selections: node.selections as any,
          };
          return transformedSelectionSet;
        },
      },
      Field: {
        leave(node) {
          const parentTypename = typeInfo.getParentType()!.name;
          const transformed: TransformedField = {
            parentTypename,
            fieldName: node.name as any,
            kind: hasListType(typeInfo.getFieldDef()!.astNode!.type)
              ? 'connection'
              : 'primitive',
            selections: node.selectionSet?.selections as any[],
          };
          return transformed;
        },
      },
      Name: {
        leave(node) {
          return node.value;
        },
      },
    })
  );
}
