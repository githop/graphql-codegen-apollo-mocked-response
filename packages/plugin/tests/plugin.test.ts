import { plugin, getConfig, Config } from '../src';
import { describe, it, expect } from 'vitest';
import { isComplexPluginOutput, Types } from '@graphql-codegen/plugin-helpers';
import { loadDocuments, loadSchema } from '@graphql-tools/load';
import { resolve } from 'path';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';

describe('tests', () => {
  describe('config', () => {
    it('sets default values with no input', () => {
      const resolvedConfig = getConfig();
      expect(resolvedConfig).toEqual({
        typesFile: './types',
        addTypename: false,
      });
    });

    it('applies overrides', () => {
      const resolvedConfig = getConfig({
        typesFile: './foo/types.ts',
        addTypename: false,
        prefix: 'b',
      });
      expect(resolvedConfig).toEqual({
        typesFile: './foo/types',
        addTypename: false,
        prefix: 'b',
      });
    });

    it('merges a partial config with defaults', () => {
      const resolvedConfig = getConfig({
        addTypename: true,
      });
      expect(resolvedConfig).toEqual({
        typesFile: './types',
        addTypename: true,
      });
    });

    it('strips file ext from typesFile', () => {
      const resolvedConfig = getConfig({ typesFile: 'foo.ts' });
      expect(resolvedConfig.typesFile).toBe('foo');
    });

    it('should honor the prefix config', async () => {
      const result = await produceExpectArg({
        prefix: 'gth',
      });

      expect(result.content).toContain('const planetMock = gthPlanet()');
      expect(result.content).toContain('const locationMock = gthLocation()');
    });

    it('should honor the types file config', async () => {
      const result = await produceExpectArg({
        typesFile: './custom-types-file.ts',
      });

      expect(result.prepend).toContain(
        "import { ListPlanetsDocument, ListPlanetsQuery, ListPlanetsQueryVariables } from './custom-types-file'"
      );
    });

    it('should honor the addTypename config', async () => {
      const result = await produceExpectArg({ addTypename: true });
      expect(result.content).toContain(
        "const ListPlanetsResult: ListPlanetsQuery = { __typename: 'Query', listPlanets: [{ __typename: 'Planet', id: planetMock.id, type: planetMock.type, mass: planetMock.mass, name: planetMock.name, location: {  __typename: 'Location', id: locationMock.id, coordinates: locationMock.coordinates, },}],}"
      );
      expect(result.content).toMatchSnapshot();
    });
  });

  describe('plugin', () => {
    it('should produce valid output', async () => {
      const mocks = await produceExpectArg();

      expect(concatPluginOutput(mocks)).toMatchSnapshot();
    });

    it('should generate type imports', async () => {
      const result = await produceExpectArg();

      expect(result.prepend).toContain(
        "import { ApolloError } from '@apollo/client';"
      );
      expect(result.prepend).toContain(
        "import { MockedResponse } from '@apollo/client/testing';"
      );
      expect(result.prepend).toContain(
        "import { ListPlanetsDocument, ListPlanetsQuery, ListPlanetsQueryVariables } from './types'"
      );
    });

    it('should include the generic interface for the mock functions', async () => {
      const result = await produceExpectArg();

      expect(result.content).toContain('interface MockFn<Variables, Query');
    });

    it('should generate a valid query result', async () => {
      const result = await produceExpectArg();

      expect(result.content).toContain(
        'const ListPlanetsResult: ListPlanetsQuery = { listPlanets: [{ id: planetMock.id, type: planetMock.type, mass: planetMock.mass, name: planetMock.name, location: {  id: locationMock.id, coordinates: locationMock.coordinates, },}],}'
      );
    });
  });
});

interface GraphQLFiles {
  schemaFilename?: string;
  documentFilename?: string;
}

async function gatherFiles({ schemaFilename, documentFilename }: GraphQLFiles) {
  const schema = await loadSchema(
    resolve(
      process.cwd(),
      '../../',
      'example',
      'src',
      'graphql',
      schemaFilename || 'schema.graphql'
    ),
    {
      loaders: [new GraphQLFileLoader()],
    }
  );

  const document = await loadDocuments(
    resolve(
      process.cwd(),
      '../../',
      'example',
      'src',
      'graphql',
      documentFilename || 'document.graphql'
    ),
    {
      loaders: [new GraphQLFileLoader()],
    }
  );

  return { schema, document };
}

function concatPluginOutput(pluginOutput: Types.PluginOutput) {
  if (isComplexPluginOutput(pluginOutput)) {
    return pluginOutput.prepend?.join('\n') + pluginOutput.content;
  }
  return pluginOutput;
}

async function produceExpectArg(
  config?: Partial<Config>,
  files: GraphQLFiles = {}
) {
  const { schema, document } = await gatherFiles(files);
  const docs = [{ location: '', document: document[0].document }];
  const result = await plugin(schema, docs, config || {});
  return result as Types.ComplexPluginOutput;
}
