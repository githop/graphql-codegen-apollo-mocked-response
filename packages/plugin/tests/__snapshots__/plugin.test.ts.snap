// Vitest Snapshot v1

exports[`tests > config > should honor the addTypename config 1`] = `
"
  interface MockFn<Input, Query> {
    ({
      input,
      result,
      error,
    }: {
      input?: Input;
      result?: Query;
      error?: ApolloError;
    }): MockedResponse<Query>;
  }  
  

  export const mockListPlanets: MockFn<
  ListPlanetsQueryVariables,
  ListPlanetsQuery
> = ({ result, input, error }) => {
  const planetMock = aPlanet();
const locationMock = aLocation();

  const ListPlanetsResult: ListPlanetsQuery = { __typename: 'Query', listPlanets: [{ __typename: 'Planet', id: planetMock.id, name: planetMock.name, location: {  __typename: 'Location', id: locationMock.id, coordinates: locationMock.coordinates, }, }], }

  return {
    request: {
      query: ListPlanetsDocument,
      variables: { input },
    },
    result: {
      data: result != null ? result : ListPlanetsResult,
      error,
    },
  };
};
  "
`;

exports[`tests > plugin > should produce valid ouput 1`] = `
"import { ApolloError } from '@apollo/client';
import { MockedResponse } from '@apollo/client/testing';
import { ListPlanetsDocument, ListPlanetsQuery, ListPlanetsQueryVariables } from './types'
  interface MockFn<Input, Query> {
    ({
      input,
      result,
      error,
    }: {
      input?: Input;
      result?: Query;
      error?: ApolloError;
    }): MockedResponse<Query>;
  }  
  

  export const mockListPlanets: MockFn<
  ListPlanetsQueryVariables,
  ListPlanetsQuery
> = ({ result, input, error }) => {
  const planetMock = aPlanet();
const locationMock = aLocation();

  const ListPlanetsResult: ListPlanetsQuery = { listPlanets: [{ id: planetMock.id, name: planetMock.name, location: {  id: locationMock.id, coordinates: locationMock.coordinates, }, }], }

  return {
    request: {
      query: ListPlanetsDocument,
      variables: { input },
    },
    result: {
      data: result != null ? result : ListPlanetsResult,
      error,
    },
  };
};
  "
`;
